### PR TITLE
Enhance the way openstack extensions are resolved.

### PR DESCRIPTION
**Pull request by zack-shoylev**
Needed for new openstack versions.

### CODE REVIEW

**Comment by zack-shoylev** [comment]
Rebuild, please.

**Comment by zack-shoylev** [comment]
Well this finally worked :)

File: apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java [hunk 1]
```diff
@@ -42,6 +43,18 @@
...
+ *
+ * 1. Match by namespace
+ * 2. Match by namespace aliases
+ * 3. Match by alias
+ * 4. Match by name
+ *
+ * New versions of openstack have no namespaces anymore.
+ * Alias is different than a namespace alias - it's an alternative namespace URL to match against.
+ *
+ * TODO: Rename PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet
```
Comment (nacx) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java]: Rename it now?

File: apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java [hunk 2]
```diff
@@ -78,7 +97,12 @@
             String arg0 = checkNotNull(args.get(0), "arg[0] in %s", input).toString();
             if (any(extensions.getUnchecked(arg0), namespaceOrAliasEquals(namespace, aliasesForNamespace)))
                return input.getResult();
-            // Could not find extension by namespace or namespace alias. Try to find it by name next:
+            // Could not find extension by namespace or namespace alias. Try to find it by alias next:
+            if (!"".equals(alias)) {
+               if (any(extensions.getUnchecked(arg0), aliasEquals(alias)))
+                  return input.getResult();
+            }
```
Comment (nacx) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java]: The logic is duplicated. Let's refactor the three lookups into a method that takes the `arg0` or the `""` as an argument and returns the extension.

**Comment by nacx** [comment]
Thx @zack-shoylev !

File: apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java (hunk 1 above)
Comment (zack-shoylev) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java]: Kinda big change. It's used in multiple APIs.

Comment (nacx) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java]: The comment indicates that you feel the name is no longer appropriate, and I agree. Usage count apart (the global rename can be mostly addresed with a search/replace), IMO it won't be a more convenient time to address this than now.

Comment (zack-shoylev) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet.java]: I looked into it more. It's not as bad as I feared. :)

**Comment by zack-shoylev** [comment]
https://github.com/jclouds/jclouds-labs-openstack/pull/203

**Comment by zack-shoylev** [comment]
Renamed and refactored, @nacx

File: apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationMatchesExtensionSet.java [hunk 3]
```diff
@@ -42,19 +43,48 @@
...
+ *
+ * 1. Match by namespace
+ * 2. Match by namespace aliases
+ * 3. Match by alias
+ * 4. Match by name
+ *
+ * New versions of openstack have no namespaces anymore.
+ * Alias is different than a namespace alias - it's an alternative namespace URL to match against.
+ *
+ * TODO: Rename PresentWhenExtensionAnnotationNamespaceEqualsAnyNamespaceInExtensionsSet
```
Comment (nacx) [inline apis/openstack-keystone/src/main/java/org/jclouds/openstack/v2_0/functions/PresentWhenExtensionAnnotationMatchesExtensionSet.java]: Remove this now :)

**Comment by nacx** [comment]
Update the `toString` method in the renamed class too.

Just these two minors. lgtm! Thanks @zack-shoylev!

**Comment by zack-shoylev** [comment]
Seems like unrelated. Hm.

**Comment by zack-shoylev** [comment]
merged

**Pull request closed without merging**